File: javascript/resistor-color-duo/resistor-color-duo.js
/**
 * Find resistor value by color code, as per
 * https://en.wikipedia.org/wiki/Electronic_color_code
 * @param {string[n]} colorBands - The array of color bands to look up (lowercase).
 * @return {integer} The corresponding resistance value.
 * @throws Will throw an error if any of the color codes are unknown (through colorCode)
 */
export const value = colorBands => {
  return colorBands
    .reverse()
    .map((band, index) => colorCode(band) * 10 ** index)
    .reduce((a, b) => a + b, 0);
};

/**
 * Find resistor value by color code, as per
 * https://en.wikipedia.org/wiki/Electronic_color_code
 * @param {string} color - The color to look up (lowercase).
 * @return {integer} The corresponding resistance value.
 * @throws Will throw an error if color code is unknown.
 */
export const colorCode = color => {
  let resistorValue = COLORS.indexOf(color);
  if (resistorValue === -1) {
    throw new Error(`Unknown resistor color: ${color}`);
  }
  return resistorValue;
};

/** The list of colors in order of increasing value */
export const COLORS = [
  "black",
  "brown",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "violet",
  "grey",
  "white"
];

File: python/resistor-color-duo/resistor_color_duo.py
"""Resistor Color Duo.
Full description is at https://exercism.org/tracks/python/exercises/resistor-color-duo.
"""

# Explicit form

RESISTOR_COLOR_MAPPING = {
    "black": 0,
    "brown": 1,
    "red": 2,
    "orange": 3,
    "yellow": 4,
    "green": 5,
    "blue": 6,
    "violet": 7,
    "grey": 8,
    "white": 9,
}


def value(colors: list[str]) -> int:
    """Calculate resistor value based on the first two color bands."""
    return RESISTOR_COLOR_MAPPING[colors[0]] * 10 + RESISTOR_COLOR_MAPPING[colors[1]]


# Implicit form

# RESISTOR_COLORS = [
#     "black",
#     "brown",
#     "red",
#     "orange",
#     "yellow",
#     "green",
#     "blue",
#     "violet",
#     "grey",
#     "white",
# ]


# def single_value(color: str) -> int:
#     """Calculate the numeric value of a single color."""
#     return RESISTOR_COLORS.index(color)


# def value(colors: list[str]) -> int:
#     """Calculate resistor value based on the first two color bands."""
#     return single_value(colors[0]) * 10 + single_value(colors[1])
